### ISSUE TITLE
Consider the performance of reporting against the cache_occurrences table

### ISSUE THREAD

**Issue by johnvanbreda** [Indicia-Team/warehouse#12]
The cache_occurrences table has grown rather large, with lots of columns required to capture most of the important fields for every record. Performance when reporting against it is not as good as we'd hope. Its been casually observed that Windows systems running anti-virus are slower than other alternatives.
The cache_occurrences table is key to reporting on Indicia so consideration of indexing and other strategies to improve performance is high priority.

**Comment by johnvanbreda** [comment]
Using a clone of warehouse1, I've picked out a couple of different queries which use cache_occurrences that were performing slowly (logged as taking 15-18 seconds). I tested the queries under pgAdmin and confirmed similar query processing times even when running them several times so PostgreSQL's cache could be as efficient as possible.
The, I created a new perftest schema. Into this I selected all the sample related columns from cache_occurrences to get all the information for each sample into a new table called perftest.cache_samples. I also created a new table called perftest.cache_occurrences with just the occurrence related columns. I excluded all the taxon dictionary columns such as default_common_name as these are already available in cache_taxa_taxon_lists. I recreated the existing indexes on the new tables.

This means that a query currently against cache_occurrences will need to be rewritten to join perftest.cache_occurrences, perftest.cache_samples and cache_taxa_taxon_lists.

I then did a very simple rewrite on the 2 test queries. In both cases the rewritten queries took approximately 5 seconds so were significantly faster.

Some further points:

* The overall volume of data on the disk would be a lot less - cache_occurrences is currently ~1590MB, the 2 new tables combined are 537MB.
* Whilst testing, I noticed a "cold start" effect where the original queries took in excess of a minute for the first run. I presumed a restart of the pg service would recreate this but it didn't - so this is just an observation and is unlikely to be Indicia related.
* These positive results were without any further consideration of indexing.
* We could mitigate against the effort required to migrate code (especially non-standard reports etc) by keeping a view called cache_occurrences which joins the 3 tables required to capture the same information.

**Comment by johnvanbreda** [comment]
For reference here are the queries to create the 2 new tables:

```sql

create schema perftest;

select distinct on (sample_id) 
  sample_id, 
  website_id, 
  survey_id, 
  survey_title, 
  date_start, 
  date_end, 
  date_type, 
  public_entered_sref, 
  entered_sref_system, 
  public_geom, 
  sample_method,
  created_by_id, 
  cache_created_on, 
  cache_updated_on,
  location_name,
  recorders,
  location_id,
  input_form,
  map_sq_1km_id,
  map_sq_2km_id,
  map_sq_10km_id,
  group_id, 
  privacy_precision,
  output_sref,
  sref_precision
into perftest.cache_samples
from cache_occurrences;

select distinct on (id)
  id, 
  record_status, 
  downloaded_flag,
  zero_abundance,
  website_id,
  sample_id,
  taxa_taxon_list_id,
  preferred_taxa_taxon_list_id,
  created_by_id,
  cache_created_on,
  cache_updated_on,
  certainty,
  verifier,
  images,
  training,
  data_cleaner_info,
  release_status,
  verified_on,
  sensitivity_precision,
  record_substatus,
  query
into perftest.cache_occurrences
from cache_occurrences;
```

**Comment by JimBacon** [comment]
The tests you mention show a great improvement in performance which is excellent. Could they be something of a special case though? What happens if you test on a query which is typically quite quick but would require a join of the two new tables?

**Comment by johnvanbreda** [comment]
Good question. I've just snapshotted a set of reports run under normal load against warehouse 1 and have started working through them to compare old vs new approaches. The results are a bit mixed - some are faster but some are slower. However, my investigations into why are hitting stumbling blocks related to how Postgres is planning the queries. To illustrate this, I've distilled an execution of the filterable_explore_list report used to populate the occurrences on the Explore page grid down to the minimum I can which shows the difference. 

Here's the query coded against the original cache_occurrences table. 
```
SELECT o.id as occurrence_id
  FROM cache_occurrences o
  WHERE o.website_id in (23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10) 
  AND o.created_by_id=3  
  ORDER BY o.id DESC 
  LIMIT 21
```

And here's the query coded against the new version:

```
SELECT o.id
  FROM perftest.cache_occurrences o
  WHERE o.website_id in (23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10)   
  AND o.created_by_id=3
 ORDER BY o.id DESC 
 LIMIT 21
```

Perftest.cache_occurrences contains an identical number of rows to (as it was cloned from) cache_occurrences. It also has exactly the same indexes applied for the columns involved in this query. It does have the columns that relate to the sample and taxon itself removed (but they are not used in this query). In terms of the disk space used as reported by PG, it is less than 10% of the size of the original. So, at the very least, performance ought to be the same, if not better. 

However here are the 2 query plans (original method first, new perftest method second):

"  ->  Sort  (cost=11543.20..11546.89 rows=1478 width=4) (actual time=9.522..9.536 rows=21 loops=1)"
"        Sort Key: id"
"        Sort Method: top-N heapsort  Memory: 25kB"
"        ->  Bitmap Heap Scan on cache_occurrences o  (cost=172.06..11503.35 rows=1478 width=4) (actual time=0.820..6.553 rows=3472 loops=1)"
"              Recheck Cond: (created_by_id = 3)"
"              Filter: (website_id = ANY ('{23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10}'::integer[]))"
"              Rows Removed by Filter: 24"
"              ->  Bitmap Index Scan on ix_cache_occurrences_created_by_id  (cost=0.00..171.69 rows=3101 width=0) (actual time=0.493..0.493 rows=3496 loops=1)"
"                    Index Cond: (created_by_id = 3)"
"Total runtime: 9.626 ms"

"Limit  (cost=0.43..2935.85 rows=21 width=4) (actual time=6.322..79.991 rows=21 loops=1)"
"  ->  Index Scan Backward using pk_perftest_cache_occurrences on cache_occurrences o  (cost=0.43..266424.77 rows=1906 width=4) (actual time=6.318..79.951 rows=21 loops=1)"
"        Filter: ((created_by_id = 3) AND (website_id = ANY ('{23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10}'::integer[])))"
"        Rows Removed by Filter: 335854"
"Total runtime: 80.042 ms"

As you can see the query plan is quite different - the original method electing to filter to created_by_id=3 first, then refilter against websites, rather than do everything in one hit. It also does a backward index scan which I thought might be the source of the problem, but if I add an index on id DESC then it switches to a forward index scan but with similar problems. I'm pretty sure the optimser's choice to scan the full index whilst filtering is basically wrong - it should use created_by_id's index to limit the rows first, then perform further filtering and sorting.

**Comment by johnvanbreda** [comment]
To summarise - I've now tested converting quite a wide range of queries now including a few different slow ones and a few different fast ones. Sticking to the ones that refer to cache_occurrences we find:

* Queries that are slow (over 5 seconds) are always significantly faster after revision at least in my tests. It appears that Postgres is much better at optimising the complex queries in the new structure.
* Queries that are fast (<1 second) are hit and miss. Some are a bit faster, some are a bit slower. In one case a query that took .3 of a second increased to approx 1 second. These differences are due to the query planner not always taking the best approach and also in some cases because the need to join across 3 tables now outstrips the other aspects of the query plan.

**Comment by DavidRoy** [comment]
I guess this now comes down revising what we are trying to achieve:
1. overall performance so therefore we need to consider query speed * how frequently run
2. reduce 'slow' performing queries so that no elements of the site are slow.  Therefore, it comes down to improving the ones that are currently slow

But either way, it seems like the revised structure is either benefitial or neutral?

**Comment by johnvanbreda** [comment]
1. I think that overall performance is not quite clear from the results of my tests. We don't have enough statistics on the prevalence of each report to judge whether one of the few queries I found that are slower are actually running very frequently, so this might have a negative impact. However we can be fairly sure that we'll reduce the chance of 2 or more big queries degrading performance by running simultaneously (which we definitely saw during the SxBRC training sessions a while back). 
2. This part I am more confident about from my tests- the noticeably slow queries are faster.

There are other implications of the change. Firstly, disk space requirements are reduced. RAM requirements are also lower because we no longer need a single monolithic cache_occurrences table loaded in RAM all the time. Also, the new structure facilitates some reporting against samples which don't have any occurrences - currently the River Barriers project for example cannot use high performance mapping tools based on map_square table mapping, since this is not directly linked to the samples table. 

As this change is quite significant in implication and I don't want to have to undo it, I wonder if we should write a small bit of code to count runs of every single type of report, so we can then double check the implications for all the frequently used reports before proceeding.

**Comment by JimBacon** [comment]
John, if I run your example query on the live warehouse the query plan is as follows.
```
Limit  (cost=0.43..12867.25 rows=21 width=4) (actual time=19.678..1679.348 rows=21 loops=1)
  Output: id
  Buffers: shared hit=76588 read=115793 dirtied=8 written=1123
  ->  Index Scan Backward using pk_cache_occurrences on indicia.cache_occurrences o  (cost=0.43..1021381.24 rows=1667 width=4) (actual time=19.673..1679.306 rows=21 loops=1)
        Output: id
        Filter: ((o.created_by_id = 3) AND (o.website_id = ANY ('{23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10}'::integer[])))
        Rows Removed by Filter: 232972
        Buffers: shared hit=76588 read=115793 dirtied=8 written=1123
Total runtime: 1679.405 ms
```
This is the same as you were seeing with your new perftest method.
I have uploaded the definition of cache_occurrences from the live server. Can we do a diff on your version of cache_occurrences to ensure they are the same. There is something going on here that doesn't seem to add up.

[cache_occurrences.txt](https://github.com/Indicia-Team/warehouse/files/28025/cache_occurrences.txt)

**Comment by johnvanbreda** [comment]
The diff of cache_occurrences on the 2 servers showed zero differences. PostgreSQL's query planning is based on many factors and therefore hard to predict - since the table stats will have changed between the live and perftest servers we might expect to see some differences in planning. 

Interestingly temporarily turning off index scans on the live warehouse when running this query results in the same plan as the fast plan on the perftest server. BUT, it is still slow - almost exactly the same speed as with index scans enabled. So, you could assume in this case that the query planner is actually picking the right plan (or at least not the wrong plan). It's just weird that on the perftest server I am seeing different behaviour, which makes me worry that it is hard to be sure the results of my tests are correct.

**Comment by JimBacon** [comment]
While the table stats may have changed, the percentage change is not going to be all that great, I wouldn't have thought.  If you 
```
vacuum (verbose, analyse) cache_occurrences
```
in both your schemas does that influence the choice of query plan?

**Comment by JimBacon** [comment]
I ran the above on the live server and didn't see any immediate change - if anything it was slower, but then the server seemed quite busy. Now I get the following.
```
Limit  (cost=12738.91..12738.97 rows=21 width=4) (actual time=27.385..27.426 rows=21 loops=1)
  Output: id
  Buffers: shared hit=301 read=2352
  ->  Sort  (cost=12738.91..12742.94 rows=1610 width=4) (actual time=27.381..27.396 rows=21 loops=1)
        Output: id
        Sort Key: o.id
        Sort Method: top-N heapsort  Memory: 25kB
        Buffers: shared hit=301 read=2352
        ->  Bitmap Heap Scan on indicia.cache_occurrences o  (cost=190.66..12695.51 rows=1610 width=4) (actual time=1.248..23.917 rows=3510 loops=1)
              Output: id
              Recheck Cond: (o.created_by_id = 3)
              Filter: (o.website_id = ANY ('{23,12,14,8,24,6,17,25,11,7,30,40,34,5,43,13,29,33,44,32,42,51,49,65,68,73,75,9,71,10}'::integer[]))
              Rows Removed by Filter: 24
              Buffers: shared hit=301 read=2352
              ->  Bitmap Index Scan on ix_cache_occurrences_created_by_id  (cost=0.00..190.25 rows=3443 width=0) (actual time=0.833..0.833 rows=3766 loops=1)
                    Index Cond: (o.created_by_id = 3)
                    Buffers: shared read=26
Total runtime: 27.509 ms
```

I don't know what is going on here but there seems to be variability of a couple of orders of magnitude!

**Comment by johnvanbreda** [comment]
I'd previously vacuumed the tables on the perftest server with no obvious effects. However this morning I ran vacuum (full, analyze) on both tables. After doing this, my test query run against the full cache_occurrences table has now switched to the Index Scan Backwards version of the plan and it runs in approx 1200ms. The exact same query plan is used against the perftest table and it takes approx 120ms, which roughly matches the size difference between the 2 tables. 

So I think the selection of the Bitmap Heap Scan backwards index plan is a bit of a fluke and we are probably exploring quirks which aren't really directly related to the structure change.

**Comment by JimBacon** [comment]
We've focussed in again on one query which may be an oddity.

In this particular instance, it seems as if, with like query plans, the revised table structure can be an order of magnitude quicker. However, if the query plan could be better chosen we could gain an increase of two orders of magnitude without having to restructure anything,

It makes evaluating the structure change more difficult and it raises another issue about whether there are more gains to be made from tuning the server.

Producing some sort of average performance test that runs a number of queries a number of times with the frequency of each query representative of normal use might do. I wonder if it would help to capture queries on the live server using log_statment set to all (http://www.postgresql.org/docs/9.1/static/runtime-config-logging.html)

**Comment by johnvanbreda** [comment]
Yes, I agree - this is an oddity. 
I've further widened my tests of different queries. Mostly, things are better in the new structure for long winded queries, and the same for queries that are relatively quick. I did find a query that runs 3 times slower (3s instead of 1s) to count the records on the NatureSpot home page but that ought to be cached anyway. 
I also found 1 particularly problematic query. This is loading the verification grid where there are a list of taxa (e.g. genera) acting as the filter. Stuart Robert's default verification filter is a good example. In the current structure, a single pass of cache_occurrences can do most of the filtering (filtering on external key, record status etc), but in the new structure we cannot filter down the cache_occurrences table much on the first pass so it all gets joined to the cache_taxa_taxon_lists table before filtering. This join then becomes a bottleneck resulting in a 1s query taking about 30s. I can fix it by moving external_key into perftest.cache_occurrences but that is a step backwards if this proposal is right. 
So, in summary although I think this change will have a positive effect in most, but not all, cases, I am not sure enough given the scale of the change. Given that the issues we have seem to relate to the size of cache_occurrences we could take a look at that aspect without a complete restructure. Therefore I'd propose an action plan as follows:

1. Review cache_occurrences and check all columns are actually being used. Remove any that aren't and consider removing any that are rarely used. Repeat for the other cache tables.
2. Reivew pg_stat_all_indexes (see http://www.postgresql.org/docs/9.3/static/monitoring-stats.html) and remove any that are not used on warehouse 1. Consider whether the space occupied by any rarely used indexes is worth it.
3. Create a report_log table and log all report accesses to it. This can have the report name, user, possibly parameters, milliseconds to run, datetime, possibly the sharing mode (so we know what they are for) and website ID. I think we ought to have this anyway. I don't think it will add much to the time required to run a report and we could always turn it off later.
4. Look at the most frequent reports, the longest running reports, and the reports with the biggest overall load (frequency * time to run). 
   * Can any of these be tidied (removal of untidy parameter filters, switch to cache tables if not already)
   * Should any of these be cached on the client side?
   * Indexing improvements possible?
5. Having made changes, repeat the capture of report_log data so that any queries we've "fixed" drop out of consideration. Then, repeat step 4 above and additionally, compare the performance of each query in the new structure.

We should then be able to get a much clearer picture of the overall change in server load/performance so we can make an informed decision, and by that time we might have made some decent improvements anyway.
